File: source/go/linked_list_reverse.go
package main

import (
	"fmt"
	"strings"
)

type Node struct {
	val  int
	next *Node
}

func buildList(arr []int) *Node {
	var head *Node
	for i := len(arr) - 1; i >= 0; i-- {
		head = &Node{val: arr[i], next: head}
	}
	return head
}

func printList(head *Node, label string) {
	var parts []string
	curr := head
	for curr != nil {
		parts = append(parts, fmt.Sprintf("%d", curr.val))
		curr = curr.next
	}
	fmt.Printf("%s: %s\n", label, strings.Join(parts, " "))
}

func reverseList(head *Node) *Node {
	var prev *Node
	curr := head
	for curr != nil {
		next := curr.next
		curr.next = prev
		prev = curr
		curr = next
	}
	return prev
}

func main() {
	head := buildList([]int{1, 2, 3, 4, 5})
	printList(head, "original")
	reversed := reverseList(head)
	printList(reversed, "reversed")
}

File: source/rust/linked_list_reverse.rs
type List = Option<Box<Node>>;

struct Node {
    val: i32,
    next: List,
}

fn build_list(arr: &[i32]) -> List {
    let mut head = None;
    for &x in arr.iter().rev() {
        head = Some(Box::new(Node { val: x, next: head }));
    }
    head
}

fn print_list(head: &List, label: &str) {
    let mut parts = Vec::new();
    let mut cur = head;
    while let Some(node) = cur {
        parts.push(node.val.to_string());
        cur = &node.next;
    }
    println!("{}: {}", label, parts.join(" "));
}

fn reverse_list(head: List) -> List {
    let mut prev: List = None;
    let mut curr = head;
    while let Some(mut node) = curr {
        curr = node.next.take();
        node.next = prev;
        prev = Some(node);
    }
    prev
}

fn main() {
    let head = build_list(&[1, 2, 3, 4, 5]);
    print_list(&head, "original");
    let reversed = reverse_list(head);
    print_list(&reversed, "reversed");
}
